### PR TITLE
Add ASGI Falcon App support

### PR DESCRIPTION
**Pull request by sarayourfriend**
This PR adds an ASGI Falcon App adapter to complement the existing Falcon adapter for WSGI apps.

I wrote this code originally based on the existing adapter for https://git.sr.ht/~sara/openverse-slack-reaction.

The code I wrote there is copied almost directly here with some modifications to raise errors when an incompatible Falcon version is being used. You can see the original code here: https://git.sr.ht/~sara/openverse-slack-reaction/tree/main/item/falcon_adapter.py

It's deployed to production with https://openverse-slack.sarayourfriend.pictures.

Hope it's okay I didn't open an issue for this first, but I needed it for my own app one way or another and copying the file  and adding tests was trivial so I didn't spend too much extra time on this compared to just writing a helpful issue describing the feature :slightly_smiling_face: 

### Category (place an `x` in each of the `[ ]`)

* [ ] `slack_bolt.App` and/or its core components
* [ ] `slack_bolt.async_app.AsyncApp` and/or its core components
* [x] Adapters in `slack_bolt.adapter`
* [ ] Document pages under `/docs`
* [ ] Others

## Requirements (place an `x` in each `[ ]`)

Please read the [Contributing guidelines](https://github.com/slackapi/bolt-python/blob/main/.github/contributing.md) and [Code of Conduct](https://slackhq.github.io/code-of-conduct) before creating this issue or pull request. By submitting, you are agreeing to those rules.

* [x] I've read and understood the [Contributing Guidelines](https://github.com/slackapi/bolt-python/blob/main/.github/contributing.md) and have done my best effort to follow them.
* [x] I've read and agree to the [Code of Conduct](https://slackhq.github.io/code-of-conduct).
* [x] I've run `./scripts/install_all_and_run_tests.sh` after making the changes.

### CODE REVIEW

File: slack_bolt/adapter/falcon/__init__.py [hunk 1]
```diff
@@ -1 +1,2 @@
 from .resource import SlackAppResource
+from .async_resource import AsyncSlackAppResource
```
Comment (seratch) [inline slack_bolt/adapter/falcon/__init__.py:2]:
```suggestion
# Don't add async module imports here
from .resource import SlackAppResource
```

File: slack_bolt/adapter/falcon/async_resource.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+    async def _to_bolt_request(self, req: Request) -> AsyncBoltRequest:
+        return AsyncBoltRequest(
+            body=(await req.stream.read(req.content_length or 0)).decode("utf-8"),
+            query=req.query_string,
+            headers={k.lower(): v for k, v in req.headers.items()},
+        )
+
+    async def _write_response(self, bolt_resp: BoltResponse, resp: Response):
+        if falcon_version.__version__.startswith("2."):
```
Comment (seratch) [inline slack_bolt/adapter/falcon/async_resource.py:59]: Can we move this validation in the constructor instead?

The if/else in the sync adapter is added to deal with the Falcon's API changes in major versions (`resp.body` in v2 vs `resp.text` in v3). As this ASGI version supports only v3, we can simply have `resp.text = bolt_resp.body` only in this adapter.

That being said, this version check is still good for better developer experience! I think that it can be done earlier than the timing that your app receive requests from Slack.

File: slack_bolt/adapter/falcon/async_resource.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+            raise NotImplementedError("ASGI Falcon requires version >= 3.0")
```
Comment (seratch) [inline slack_bolt/adapter/falcon/async_resource.py:60]:
```suggestion
            raise BoltError("This ASGI compatible adapter requires Falcon version >= 3.0")
```

File: tests/adapter_tests_async/test_async_falcon.py [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+from tests.mock_web_api_server import (
+    setup_mock_web_api_server,
+    cleanup_mock_web_api_server,
+    assert_auth_test_count,
+)
+from tests.utils import remove_os_env_temporarily, restore_os_env
+
+
+def new_falcon_app():
+    if falcon.version.__version__.startswith("2."):
```
Comment (seratch) [inline tests/adapter_tests_async/test_async_falcon.py:23]: You can remove this if condition

**Review by seratch — CHANGES_REQUESTED** [review]
Thanks for the great addition! We are happy to have this enhancement but could you check my comments and update your branch?

File: slack_bolt/adapter/falcon/__init__.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sarayourfriend) [reply slack_bolt/adapter/falcon/__init__.py:2]: Interesting, that hadn't crossed my mind, thanks for catching it!

[the remark this answers is not in the document]
Comment (sarayourfriend) [reply slack_bolt/adapter/falcon/__init__.py]: Done in https://github.com/slackapi/bolt-python/pull/614/commits/937fddc20c28c2f467a71f1a849b4287b0ab362d

File: slack_bolt/adapter/falcon/async_resource.py (hunk 2 above)
Comment (sarayourfriend) [reply slack_bolt/adapter/falcon/async_resource.py]: Done in https://github.com/slackapi/bolt-python/pull/614/commits/515345c42aa1ca12034650f5d231829d4d27adf5

File: tests/adapter_tests_async/test_async_falcon.py (hunk 4 above)
Comment (sarayourfriend) [reply tests/adapter_tests_async/test_async_falcon.py]: Done in https://github.com/slackapi/bolt-python/pull/614/commits/515345c42aa1ca12034650f5d231829d4d27adf5

**Comment by sarayourfriend** [comment]
Thanks for the quick review @seratch! I've updated the PR with your requested changes.

File: examples/falcon/async_app.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+
+api = falcon.asgi.App()
+resource = AsyncSlackAppResource(app)
+api.add_route("/slack/events", resource)
+
+# pip install -r requirements.txt
+# export SLACK_SIGNING_SECRET=***
+# export SLACK_BOT_TOKEN=xoxb-***
+# uvicorn --reload -h 0.0.0.0 -p 3000 app:api
```
Comment (seratch) [inline examples/falcon/async_app.py:100]:
```suggestion
# uvicorn --reload -h 0.0.0.0 -p 3000 async_app:api
```

File: examples/falcon/async_oauth_app.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+
+api = falcon.asgi.App()
+resource = AsyncSlackAppResource(app)
+api.add_route("/slack/events", resource)
+
+# pip install -r requirements.txt
+# export SLACK_SIGNING_SECRET=***
+# export SLACK_BOT_TOKEN=xoxb-***
+# uvicorn --reload -h 0.0.0.0 -p 3000 app:api
```
Comment (seratch) [inline examples/falcon/async_oauth_app.py:100]:
```suggestion
# uvicorn --reload -h 0.0.0.0 -p 3000 async_oauth_app:api
```

**Review by seratch — CHANGES_REQUESTED** [review]
Can you update a few comments in the example apps?

File: examples/falcon/async_app.py (hunk 5 above)
Comment (seratch) [reply examples/falcon/async_app.py]: @sarayourfriend Also, I would suggest gunicorn here because the requirements.txt has only gunicorn

**Comment by seratch** [comment]
@sarayourfriend Thanks for quick updates on the feedback! My last request is to update the examples a bit. Once the builds pass, we can merge this PR and release a new version by the end of this week.

**Comment by sarayourfriend** [comment]
Hmm, it seems like there's an issue with the fact that we have tests that expect Falcon 2 instead of 3... is there some way to exclude the async Falcon tests when the falcon version is `2.x`?

File: examples/falcon/async_app.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+
+
+api = falcon.asgi.App()
+resource = AsyncSlackAppResource(app)
+api.add_route("/slack/events", resource)
+
+# pip install -r requirements.txt
+# export SLACK_SIGNING_SECRET=***
+# export SLACK_BOT_TOKEN=xoxb-***
+# uvicorn --reload -h 0.0.0.0 -p 3000 async_app:api
```
Comment (seratch) [inline examples/falcon/async_app.py:100]: @sarayourfriend Can you add unvicorn to requirements.txt in the same directory?

**Comment by seratch** [comment]
@sarayourfriend You can change this line to use `>=3,<4` instead: https://github.com/slackapi/bolt-python/blob/v1.11.6/setup.py#L82
Changing the part does not affect users.

**Comment by sarayourfriend** [comment]
@seratch should I also remove the conditional in the existing Falcon adapter tests? I've never worked with `setup.py` before so just want to make sure I'm understanding correctly that it will run the tests only with Falcon 3.x?

It could make this function unnecessary: https://github.com/sarayourfriend/bolt-python/blob/78b64878b692039e39530888a012bf738ca74058/tests/adapter_tests/falcon/test_falcon.py#L23

File: examples/falcon/async_app.py (hunk 7 above)
Comment (sarayourfriend) [reply examples/falcon/async_app.py:100]: Sure thing. I'm not sure what version range to pin it to. The project is still <1. Should I just use `uvicorn>=0.17.6` (the currently released version)?

**Comment by seratch** [comment]
@sarayourfriend Ah, sorry - never mind my above comment. I totally forget what I did in the past 🤦 No need to modify `setup.py`. 

Can you modify https://github.com/slackapi/bolt-python/blob/v1.11.6/.github/workflows/tests.yml#L73 to have `pip install "falcon>=3,<4"` before the pytest execution? It should resolve the issue.

**Comment by sarayourfriend** [comment]
@seratch just wanted to make sure, because I don't see any other explicit dependency installations in that workflow... would it make sense to add the `falcon>=3` range in the `async` dependency list as that's already being installed in the `tests.yml`

This is the patch I would apply if that makes sense:

```diff
diff --git a/setup.py b/setup.py
index 6543529..9c92d94 100755
--- a/setup.py
+++ b/setup.py
@@ -68,6 +68,8 @@ setuptools.setup(
             "aiohttp>=3,<4",
             # Socket Mode 3rd party implementation
             "websockets>=8,<10",
+            # Falcon introduced ASGI support in v3
+            "falcon>=3,<4",
         ],
         # pip install -e ".[adapter]"
         # NOTE: any of async ones requires pip install -e ".[async]" too
@@ -79,7 +81,7 @@ setuptools.setup(
             "click>=7,<8",  # for chalice
             "CherryPy>=18,<19",
             "Django>=3,<5",
-            "falcon>=3,<4",
+            "falcon>=2,<4",
             "fastapi>=0.70.0,<1",
             "Flask>=1,<3",
             "Werkzeug>=2,<3",
```

If this doesn't work for some reason or you still prefer the explicit installation in the `tests.yml` script then let me know and I'll make the appropriate change.

**Comment by seratch** [comment]
@sarayourfriend this line manually installs v2: https://github.com/slackapi/bolt-python/blob/v1.11.6/.github/workflows/tests.yml#L56 this is necessary for v2 tests but it causes errors in the async tests you added.

File: examples/falcon/async_app.py (hunk 7 above)
Comment (seratch) [reply examples/falcon/async_app.py:100]: You can do the same with fastapi examples

**Comment by sarayourfriend** [comment]
Ah, I see. Makes sense, thanks for explaining. Pushed the update!

File: setup.py [hunk 8]
```diff
@@ -79,7 +79,7 @@
             "click>=7,<8",  # for chalice
             "CherryPy>=18,<19",
             "Django>=3,<5",
-            "falcon>=2,<4",
+            "falcon>=3,<4",
```
Comment (seratch) [inline setup.py:82]: Can you revert this change? This is not necessary
```suggestion
            "falcon>=2,<4",
```

File: slack_bolt/adapter/falcon/async_resource.py [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+
+    async def _to_bolt_request(self, req: Request) -> AsyncBoltRequest:
+        return AsyncBoltRequest(
+            body=(await req.stream.read(req.content_length or 0)).decode("utf-8"),
+            query=req.query_string,
+            headers={k.lower(): v for k, v in req.headers.items()},
+        )
+
+    async def _write_response(self, bolt_resp: BoltResponse, resp: Response):
+        status = HTTPStatus(bolt_resp.status)
```
Comment (seratch) [inline slack_bolt/adapter/falcon/async_resource.py:63]: As you completely removed the part setting response body here, the unit test is correctly failing.
```suggestion
        resp.text = bolt_resp.body
        status = HTTPStatus(bolt_resp.status)
```

**Review by seratch — CHANGES_REQUESTED** [review]
@sarayourfriend I found that your unit test properly detects a bug in your code. Can you check my comments?

File: slack_bolt/adapter/falcon/async_resource.py (hunk 9 above)
Comment (sarayourfriend) [reply slack_bolt/adapter/falcon/async_resource.py:63]: Oh dear :cold_sweat: I was not careful in that change. Thanks for finding the fix.

File: setup.py (hunk 8 above)
Comment (sarayourfriend) [reply setup.py:82]: I wondered about this! :+1:

**Review by seratch — APPROVED** [review]
LGTM
